### PR TITLE
HPCC-28128 First set of improvements to cse code

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
This is submitted as 4 separate commits, because each one is independent and separately tested.
All but the last have no effect on the generated code - they are fixing issues that appeared when other changes were made.  The last change improves the generated code for a set of queries.  One (tkirk29) gets slightly worse - but not enough to warrant any concerns.

**Comment by ghalliday** [comment]
(Compare ignoring whitespace)

**Comment by ghalliday** [comment]
Actually I remembered incorrectly, commit 1 allowed a few if activities to be combined into a case activity.  Commit 3 is the one that caused some significant improvements.

File: ecl/hqlcpp/hqlcse.cpp [hunk 1]
```diff
@@ -292,26 +296,40 @@ void CseSpotter::analyseExpr(IHqlExpression * expr)
         extra->canAlias = true;
 
     bool savedCanAlias = canAlias;
-    if (expr->isDataset() && (op != no_select) && (!spotCseInIfDatasetConditions || (op != no_if)))
+    bool processed = false;
+    if (canAlias)
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlcse.cpp:300]: note: This is primarily restructuring the code.  The only actual difference is for no_createset - which prevents expressions from the underlying dataset from being aliased (which caused problems in combination of some other changes)

**Review by shamser — APPROVED** [review]

**Review by richardkchapman — APPROVED** [review]

**Comment by richardkchapman** [comment]
Unit tests are reporting a core - not sure if related but should be checked.

**Comment by ghalliday** [comment]
That was a path issue - the test crashes if it cannot find the libloggingmanager.so (which happened when I ran it from the build directory)

**Comment by ghalliday** [comment]
@richardkchapman unit test now passing (I wonder if it was the roxie file problem).  The extra commit to improve the unit test should be harmless/beneficial.

**Pull request merged**